### ISSUE TITLE
Repeated invocations with `--noallow_analysis_cache_discard` don't fail as expected

### ISSUE THREAD

**Issue by lberki** [bazelbuild/bazel#23491]
### Description of the bug:

If I run a build with `--noallow_analysis_cache_discard` and with changed configuration flags, it fails as expected.

However, if I run the same build a second time, it runs, thereby discarding the analysis cache despite being asked not to do so.

### Which category does this issue belong to?

Configurability

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
cat > BUILD <<'EOF'
cc_binary(name="cc", srcs=["cc.cc"])
EOF

cat > cc.cc << 'EOF'
int main(void) {
  return 0;
}
EOF

bazel build -c opt :cc
bazel build --noallow_analysis_cache_discard -c dbg :cc  # Fails
bazel build --noallow_analysis_cache_discard -c dbg :cc  # Builds, but it should not (and discards the analysis cache)
```

Furthermore, switching *back* to the original configuration fails, although it work:

```
bazel build -c opt :cc
bazel build --noallow_analysis_cache_discard -c dbg :cc  # Fails
bazel build --noallow_analysis_cache_discard -c opt :cc  # Fails, although analysis cache has -c opt
````

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

Google-internal version

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

Used the currently released version at Google.

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.

My guess would be https://github.com/bazelbuild/bazel/pull/16805, the initial implementation of this flag although I haven't dug into it.

**Comment by gregestren** [comment]
Fix incoming at https://bazel-review.googlesource.com/c/bazel/+/258477 (by @lberki ).

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0

**Comment by lberki** [comment]
Manually merged pull request is https://github.com/bazelbuild/bazel/pull/23653 .

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
